zerver/models.py

```diff
@@ -1385,6 +1386,19 @@ def remote_user_to_email(remote_user: str) -> str:
 # whenever we save it.
 post_save.connect(flush_user_profile, sender=UserProfile)
 
+def adjust_user_subscriptions_if_needed(sender: Any, **kwargs: Any) -> None:
+    user_profile = kwargs['instance']
+
+    if changed(kwargs, ['is_active']):
+        active_status = user_profile.is_active
+        Subscription.objects.filter(user_profile=user_profile).update(is_user_active=active_status)
+
+post_save.connect(
+    adjust_user_subscriptions_if_needed,
+    sender=UserProfile,
+    dispatch_uid="adjust_user_subscriptions_if_needed"
+)
```

Comment:
Just to check, are we guaranteed that this `post_save` hook runs inside the same transaction as the deactivation/reactivation?  

If not, this is probably the wrong design; we can always do a `with transaction.atomic` in `do_deactivate_user` and friends.

Reply:
Yeah, atomicity wouldn't be guaranteed. `change_user_is_active` is a new weird function that can be used to get this done atomically. I've kept the signal handler which is why the function is kind of weird - I'm not sure if this is the right approach or we should just remove the signal handler and do the Subscriptions update query in the atomic block in `change_user_is_active` directly?

Keeping the signal handler has the small upside that it'll still work if some code changes `user.is_active` directly without properly calling change_user_is_active - but perhaps we should just add a semgrep rule to prevent such a thing?

Reply:
I got rid of the signal, in favor of doing the Subscription update directly in the helper function + a semgrep rule to protect from new code mutating .is_active directly

Reply:
Yeah, this makes sense to me.